### DISCUSSION TITLE
Creating FLAT index did not call knowhere::VecIndexFactory::CreateVecIndex?

### DISCUSSION THREAD

**Discussion by crazyhank** [milvus-io/milvus#26001] [Q&A and General discussion]
I use hello_milvus.py to test, lool at the following code:
`
index = {
    "index_type": "IVF_FLAT",
    "metric_type": "L2",
    "params": {"nlist": 128},
}

hello_milvus.create_index("embeddings", index)
`
Here, when I use IVF_FLAT type, I found the milvus will call knowhere::VecIndexFactory::CreateVecIndex function to create a real faiss IVF_FLAT index, but when I change the index type to FLAT, I thought it will still call this funtion, but it did not. As I exepect, it will create a real faiss IDMAP index, am I wrong, correct me please, thanks.

**Comment by crazyhank** [comment]
It seems that it will call BruteForce interface in knowhere if I set FLAT index type. How can I let mivus call Faiss IndexIDMAP if I want to use FLAT index?

**Comment by crazyhank** [comment]
I found the following code which maybe is the reason:
`
if isFlatIndex(getIndexType(indexParams)) || meta.NumRows < Params.IndexCoordCfg.MinSegmentNumRowsToEnableIndex {
			log.Ctx(ib.ctx).Debug("segment does not need index really", zap.Int64("buildID", buildID),
				zap.Int64("segID", meta.SegmentID), zap.Int64("num rows", meta.NumRows))
`
So, if index type specified by user is FLAT, it will automatically use brute force search, then how can I use Faiss GPU search which is much faster than CPU.

**Reply by yhmo** [reply]
In Milvus, data is split into segments. For each segment, Milvus builds an independent index for it.
Yes, FLAT is a brute-force search "index". For segments that contain rows less than 1024, Milvus will skip building index for them. So, small segments are always FLAT index no matter what index the user specified.

All the versions v2.2.x **cannot** support GPU index. Only the incoming v2.3 can support GPU index. Defferent index types are defined for GPU:  GPU_FLAT and GPU_IVF_FLAT. The v2.3-beta is released one month ago, but it is not stable. We are going to release an official v2.3 in one or two weeks.

**Reply by crazyhank** [reply]
Thanks for your reply. I did a test to remove skip building index code in index_builder.go file and rebuild it. It seems that it worked well now and as expected it will call Faiss IndexIDMAP interface.
Actucally, we are now adding our hardware to Milvus  index backend, because it has much faster speed than CPU when doing brutal force search. Hope that it will work, :)
